dialogue-core/src/main/java/com/palantir/dialogue/core/RetryingChannel.java

```diff
@@ -37,46 +48,87 @@
+        this(delegate, maxRetries, backoffSlotSize, serverQoS, sharedScheduler.get(), () ->
+                ThreadLocalRandom.current().nextDouble());
+    }
+
+    RetryingChannel(
+            Channel delegate,
+            int maxRetries,
+            Duration backoffSlotSize,
+            ClientConfiguration.ServerQoS serverQoS,
+            ListeningScheduledExecutorService scheduler,
+            DoubleSupplier random) {
         this.delegate = delegate;
         this.maxRetries = maxRetries;
+        this.backoffSlotSize = backoffSlotSize;
         this.serverQoS = serverQoS;
+        this.scheduler = scheduler;
+        this.random = random;
     }
 
     @Override
     public ListenableFuture<Response> execute(Endpoint endpoint, Request request) {
-        return new RetryingCallback(delegate, endpoint, request, maxRetries, serverQoS).execute();
+        return new RetryingCallback(endpoint, request).execute();
     }
 
-    private static final class RetryingCallback {
-        private final Channel delegate;
+    private final class RetryingCallback {
         private final Endpoint endpoint;
         private final Request request;
-        private final int maxRetries;
-        private final ClientConfiguration.ServerQoS serverQoS;
         private int failures = 0;
 
-        private RetryingCallback(
-                Channel delegate,
-                Endpoint endpoint,
-                Request request,
-                int maxRetries,
-                ClientConfiguration.ServerQoS serverQoS) {
-            this.delegate = delegate;
+        private RetryingCallback(Endpoint endpoint, Request request) {
             this.endpoint = endpoint;
             this.request = request;
-            this.maxRetries = maxRetries;
-            this.serverQoS = serverQoS;
         }
 
         ListenableFuture<Response> execute() {
             return wrap(delegate.execute(endpoint, request));
         }
 
+        @SuppressWarnings("FutureReturnValueIgnored") // error-prone bug
+        ListenableFuture<Response> retry(Throwable cause) {
+            long backoffNanoseconds = getBackoffNanoseconds();
+            logRetry(backoffNanoseconds, cause);
+            if (backoffNanoseconds <= 0) {
+                return wrap(delegate.execute(endpoint, request));
+            }
+            ListenableScheduledFuture<ListenableFuture<Response>> scheduled = scheduler.schedule(
+                    () -> delegate.execute(endpoint, request), backoffNanoseconds, TimeUnit.NANOSECONDS);
+            return wrap(Futures.transformAsync(scheduled, input -> input, MoreExecutors.directExecutor()));
+        }
+
+        private long getBackoffNanoseconds() {
+            if (failures == 0) {
+                return 0L;
+            }
+            int upperBound = (int) Math.pow(2, failures - 1);
+            return Math.round(backoffSlotSize.toNanos() * random.getAsDouble() * upperBound);
```

Comment:
We're gonna plumb the BackoffStrategy thing in here right?

Reply:
We could, but I don't like it as written in CJR, I find it confusing that getting the duration mutates state and increments failures. I had an implementation that took a backoffstrategy as a function from failure count to backoff duration that worked fine, but without needing alternative implementations it didn't add anything.
Thoughts?

Reply:
Yeah definitely would like to not do the mutation as part of the getter. Fair enough that if we're always going to use the exponential strategy, then maybe factoring out is unnecessary!